package/MDAnalysis/coordinates/TNG.py

```diff
@@ -0,0 +1,437 @@
+        self._file_iterator._open(self.filename, 'r')
+
+    def _frame_to_step(self, frame: int) -> int:
+        """ Convert a frame number to an integrator step
+
+        Parameters
+        ----------
+        frame : int
+            The frame number
+
+        Returns
+        -------
+        integrators_step :  int
+            The integrator step of the frame in the TNG file
+
+        """
+        return frame * self._global_stride
+
+    def _read_frame(self, i: int) -> Timestep:
+        """read frame i
+
+        Parameters
+        ----------
+        i : int
+            The trajectory frame to be read
+
+        Returns
+        -------
+        ts : Timestep
+            Data from frame i encapsulated in an MDA `:class:Timestep` 
+        """
+        self._frame = i - 1
+        ts = self._read_next_timestep()
+        return ts
+
+    def _read_next_timestep(self, ts: Optional[Timestep] = None) -> Timestep:
+        """Read next frame into a timestep
+
+        Parameters
+        ----------
+        ts : Timestep
+            The timestep to read the data into
+
+        Returns
+        -------
+        ts :  Timestep
+            The timestep filled with data from the next step        
+        """
+        if self._frame == self.n_frames - 1:
+            raise IOError('trying to go over trajectory limit')
+        if ts is None:
+            ts = self.ts
+        # convert from frames to integrator steps
+        step = self._frame_to_step(self._frame + 1)
+        iterator_step = self._file_iterator.read_step(step)
+        self._frame += 1
+        ts = self._frame_to_ts(iterator_step, ts)
+        return ts
+
+    
+    def _frame_to_ts(self, curr_step: 'pytng.TNGCurrentIntegratorStep',
+                     ts: Timestep) -> Timestep:
+        """convert a TNGCurrentIteratorStep to an MDA Timestep
+
+        Parameters
+        ----------
+        curr_step : pytng.TNGCurrentIntegratorStep
+            The current timestep in the TNG trajectory
+
+        ts : Timestep
+            The current timestep in the MDA trajectory
+
+        Returns
+        -------
+        ts: Timestep
+            The current timestep in the MDA trajectory with data from TNG
+            trajectory integrator step
+
+        Raises
+        ------
+        IOError
+            The reading of one of the attributes from the TNG file failed
+        """
+
+        ts.frame = self._frame
+        ts.time = curr_step.get_time()
+        ts.data['step'] = curr_step.step
+
+        if self._has_box:
+            curr_step.get_box(self._box_temp)
+            ts.dimensions = triclinic_box(*self._box_temp.reshape(3, 3))
+            if not curr_step.read_success:
+                raise IOError("Failed to read box from TNG file")
+        if self._has_positions:
+            curr_step.get_positions(ts.positions)
+            if not curr_step.read_success:
+                raise IOError("Failed to read positions from TNG file")
```

Comment:
I've not read the pytng code, but why are these errors not happening in pytng itself?

Reply:
PyTNG is designed to mimic the TNG API and not complain too much if you try and read where there isn't data. This often occurs when some blocks are present and not others,  but I guess I could consider shifting the testing "strictness" downstream if you think that is cleaner?